atomicapp/providers/lib/kubeshift/openshift.py

```diff
@@ -0,0 +1,384 @@
+            replicates (int): Default 0, size of the amount of replicas to scale
+        '''
+        patch = [{"op": "replace",
+                  "path": "/spec/replicas",
+                  "value": replicas}]
+        name = self._get_metadata_name(obj)
+        _, url = self._generate_kurl(obj, namespace, name)
+        self.api.request("patch", url, data=patch)
+        logger.info("'%s' successfully scaled to %s", name, replicas)
+
+    def namespaces(self):
+        '''
+        Gathers a list of namespaces on the Kubernetes cluster
+        '''
+        url = urljoin(self.k8s_api, "namespaces")
+        ns = self.api.request("get", url)
+        return ns['items']
+
+    def _generate_kurl(self, obj, namespace, name=None, params=None):
+        '''
+        Generate the required URL by extracting the 'kind' from the
+        object as well as the namespace.
+
+        Args:
+            obj (obj): Object of the data being passed
+            namespace (str): k8s namespace
+            name (str): Name of the object being passed
+            params (arr): Extra params passed such as timeout=300
+
+        Returns:
+            kind (str): The kind used
+            url (str): The URL to be used / artifact URL
+        '''
+        if 'apiVersion' not in obj.keys():
+            raise KubeOpenshiftError("Error processing object. There is no apiVersion")
+
+        if 'kind' not in obj.keys():
+            raise KubeOpenshiftError("Error processing object. There is no kind")
+
+        api_version = obj['apiVersion']
+
+        kind = obj['kind']
+
+        resource = KubeBase.kind_to_resource_name(kind)
+
+        if resource in self.k8s_api_resources[api_version]:
+            if api_version == 'v1':
+                url = self.k8s_api
+            else:
+                url = urljoin(self.k8s_apis, "%s/" % api_version)
+        else:
+            raise KubeOpenshiftError("No kind by that name: %s" % kind)
+
+        url = urljoin(url, "namespaces/%s/%s/" % (namespace, resource))
+
+        if name:
+            url = urljoin(url, name)
+
+        if params:
+            url = urljoin(url, "?%s" % urlencode(params))
+
+        return (resource, url)
+
+    @staticmethod
+    def _get_metadata_name(obj):
+        '''
+        This looks at the object and grabs the metadata name of said object
+
+        Args:
+            obj (object): Object file of the artifact
+
+        Returns:
+            name (str): Returns the metadata name of the object
+        '''
+        if "metadata" in obj and \
+                "name" in obj["metadata"]:
+            name = obj["metadata"]["name"]
+        else:
+            raise KubeOpenshiftError("Cannot undeploy. There is no"
+                                     " name in object metadata "
+                                     "object=%s" % obj)
+        return name
+
+    # OPENSHIFT-SPECIFIC FUNCTIONS
+
+    def extract(self, image, src, dest, namespace, update=True):
```

Comment:
Yeah... I haven't tested this with Kubernetes yet. This was added since this is atomicapp-specific code. I'm considering removing it all together and putting it in the `providers/openshift.py` code since it doesn't really "fit" in here. Seems more of a hack. Those are my thoughts.